Implementation of lp:per-session-server-id bluepring

gl-sergei opened the pull request:
Added session variable pseudo_server_id with default value 0. When
pseudo_server_id is not zero, binlog events get logged with server_id
= pseudo_server_id.

Changing pseudo_server_id requires SUPER privilege.

Code change is trivial. Two MTR test added testing simple
master-slave setup and circular replication setup.

gl-sergei commented:
http://jenkins.percona.com/view/PS%205.6/job/percona-server-5.6-param/919/

laurynas-biveinis commented on mysql-test/suite/sys_vars/r/all_vars.result:
```diff
@@ -16,5 +16,6 @@ INNODB_FAKE_CHANGES
 INNODB_FAKE_CHANGES
 INNODB_KILL_IDLE_TRANSACTION
 INNODB_KILL_IDLE_TRANSACTION
+PSEUDO_SERVER_ID
```
Why not a pseudo_server_id_basic.test ?

laurynas-biveinis commented on mysql-test/suite/rpl/t/pseudo_server_id.test:
```diff
@@ -0,0 +1,100 @@
+connect(con_user1,localhost,user1,"111",,);
+--connection con_user1
+select @@session.pseudo_server_id;
+--error 1227
```
Please use ER_FOO instead of a number

laurynas-biveinis commented:
Looks good, just how about testing:
- setting pseudo_server_id and then server_id in the same thread;
- --replicate-same-server-id

I'd also considering something similar to mysql-test/include/show_binlog_events.inc, except that it wouldn't mask server_id, and adding it to the existing testcases.

gl-sergei commented:
1. Added parameter `keep_server_id` for `show_binlog_events.inc` and added `show_binlog_events.inc` to existing tests.
2. Moved basic test and access rights test to `pseudo_server_id_basic` test
3. Added test setting `server_id` and `pseudo_server_id` in the same session
4. Added test with `replicate-same-server-id`

http://jenkins.percona.com/view/PS%205.6/job/percona-server-5.6-param/932/
